libuavcan/include/libuavcan/transport/media/can.hpp

```diff
@@ -0,0 +1,365 @@
+}  // end namespace TypeFD
+
+/**
+ * @namespace Type2_0
+ * Properties of a CAN-2.0 compliant bus.
+ */
+namespace Type2_0
+{
+/**
+ * The maximum size of a data frame for this bus.
+ */
+constexpr static std::size_t MaxFrameSizeBytes = 8;
+
+/**
+ * Lookup table to find the data length that would be used to
+ * store a given payload.
+ */
+constexpr static std::array<std::uint8_t, MaxFrameSizeBytes> PayloadLengthToFrameLength = {{0, 1, 2, 3, 4, 5, 6, 7}};
+
+}  // end namespace Type2_0
+
+/**
+ * Bit pattern to fill padding bytes with. The UAVCAN specification mandates this value
+ * but the actual value of padding bytes must be ignored when receiving messages.
+ *
+ * When transmitting use this pattern to minimize the number of stuff bits added by the
+ * CAN hardware.
+ */
+constexpr std::uint8_t BytePaddingPattern = 0x55;
+
+/**
+ * Valid message data codes.
+ *
+ * CAN DLCs are only 4-bits long so FD had to use 9-15 to encode
+ * various lengths up to 64 bits.
+ */
+enum class FrameDLC : std::uint_fast8_t
+{
+    CodeForLength0  = 0,  /**< Data Length Code: 0 bytes */
+    CodeForLength1  = 1,  /**< Data Length Code: 1 bytes */
+    CodeForLength2  = 2,  /**< Data Length Code: 2 bytes */
+    CodeForLength3  = 3,  /**< Data Length Code: 3 bytes */
+    CodeForLength4  = 4,  /**< Data Length Code: 4 bytes */
+    CodeForLength5  = 5,  /**< Data Length Code: 5 bytes */
+    CodeForLength6  = 6,  /**< Data Length Code: 6 bytes */
+    CodeForLength7  = 7,  /**< Data Length Code: 7 bytes */
+    CodeForLength8  = 8,  /**< Data Length Code: 8 bytes */
+    CodeForLength12 = 9,  /**< Data Length Code: 12 bytes */
+    CodeForLength16 = 10, /**< Data Length Code: 16 bytes */
+    CodeForLength20 = 11, /**< Data Length Code: 20 bytes */
+    CodeForLength24 = 12, /**< Data Length Code: 24 bytes */
+    CodeForLength32 = 13, /**< Data Length Code: 32 bytes */
+    CodeForLength48 = 14, /**< Data Length Code: 48 bytes */
+    CodeForLength64 = 15, /**< Data Length Code: 64 bytes */
+};
+
+/**
+ * Raw CAN frame, as passed to/from a CAN driver.
+ *
+ * @tparam  MTUBytesParam       The maximum number of bytes that can be stored in this frame.
+ * @tparam  FlagBitsCompareMask A mask of the upper thee bits of this class's id field. By
+ *                              default these bits are ignored by equality operators in the
+ *                              class. If an implemenation wants to use these three bits as
+ *                              meta-data it should provide a different compare mask to include
+ *                              the bits when comparing identifiers.
+ */
+template <std::uint16_t MTUBytesParam, std::uint8_t FlagBitsCompareMask = 0x00>
+struct LIBUAVCAN_EXPORT Frame
+{
+    static_assert(MTUBytesParam <= TypeFD::MaxFrameSizeBytes, "CAN::Frame cannot hold anything larger than an CAN FD frame.");
+
+    /**
+     * 29-bit mask for extended frame identifiers.
+     */
+    static constexpr std::uint32_t MaskExtID = 0x1FFFFFFFU;
```

Comment:
Do we still need this, considering that we only support 29-bit ID now?

Reply:
Yes. This protects the user from comparing junk bits since we do not define what happens to the 3 MSB in uint32. See the FlagBitsCompareMask for how we make these bits available to the user.

I'm about to post one more change that enhances the documentation for this and for the Frame class in general...